src/submitter.py

```diff
@@ -0,0 +1,405 @@
+    if access_key and secret_access_key:
+        session_kwargs['aws_access_key_id'] = access_key
+        session_kwargs['aws_secret_access_key'] = secret_access_key
+    session = boto3.session.Session(**session_kwargs)
+
+    kwargs = {
+        'service_name': 's3',
+        'config': Boto3Config(s3={'addressing_style': 'path'}),
+    }
+    if region_name:
+        kwargs['region_name'] = region_name
+
+    if endpoint_url:
+        kwargs['endpoint_url'] = endpoint_url
+
+    return session.client(**kwargs)
+
+
+def s3_fetch(client, bucket, key):
+    """Fetches a key from S3
+
+    :arg client: S3 client
+    :arg bucket: S3 bucket name
+    :arg key: key for the item to fetch
+
+    :returns: item as bytes
+
+    """
+    data = io.BytesIO()
+    client.download_fileobj(bucket, key, data)
+    return data.getvalue()
+
+
+def generate_s3_key(kind, crash_id):
+    """Generates the key in S3 for this object kind
+
+    :arg kind: the kind of thing to fetch
+    :arg crash_id: the crash id
+
+    :returns: the key name
+
+    """
+    if kind == 'raw_crash':
+        return 'v2/raw_crash/%s/20%s/%s' % (crash_id[0:3], crash_id[-6:], crash_id)
+    if kind == 'dump_names':
+        return 'v1/dump_names/%s' % crash_id
+    if kind in (None, '', 'upload_file_minidump'):
+        kind = 'dump'
+    return 'v1/%s/%s' % (kind, crash_id)
+
+
+def fetch_raw_crash(client, bucket, crash_id):
+    """Fetches raw crash and converts from JSON to Python dict"""
+    key = generate_s3_key('raw_crash', crash_id)
+    data = s3_fetch(client, bucket, key).decode('utf-8')
+    return json.loads(data)
+
+
+def fetch_dumps(client, bucket, crash_id):
+    """Fetches dump data and returns list of (name, data) tuples"""
+    dumps = {}
+
+    # fetch dump_names
+    key = generate_s3_key('dump_names', crash_id)
+    dump_names = json.loads(s3_fetch(client, bucket, key))
+
+    # fetch dumps
+    for name in dump_names:
+        key = generate_s3_key(name, crash_id)
+        dumps[name] = s3_fetch(client, bucket, key)
+
+    return dumps
+
+
+def smart_bytes(thing):
+    if isinstance(thing, bytes):
+        return thing
+
+    if isinstance(thing, (float, int)):
+        return str(thing).encode('utf-8')
+
+    return thing.encode('utf-8')
+
+
+def multipart_encode(raw_crash, dumps):
+    """Takes a raw_crash and list of (name, dump) and converts to a multipart/form-data
+
+    This returns a tuple of two things:
+
+    1. a ``bytes`` object with the HTTP POST payload
+    2. a dict of headers with ``Content-Type`` and ``Content-Length`` in it
+
+    :arg payload_dict: Python dict of name -> value pairs. Values must be either
+         strings or a tuple of (filename, file-like objects with ``.read()``).
+
+    :returns: tuple of (bytes, headers dict)
+
+    """
+    # NOTE(willkg): This is the result of uuid.uuid4().hex. We just need a
+    # unique string to denote the boundary between parts in the payload.
+    boundary = '01659896d5dc42cabd7f3d8a3dcdd3bb'
```

Comment:
Having this be stable and not randomly generated vastly simplifies testing payload generation.